doc/arch/020-questionnaires-aka-dynamic-forms.md

```diff
@@ -0,0 +1,127 @@
+- **Layout:** A custom DSL that helps Mobile rendering a form defined by server.
+
+## Design (AKA Implementation Details)
+
+Questionnaires are implemented by following below mentioned steps:
+
+### 1. Mobile _syncs_ questionnaires from server to device
+Mobile requests questionnaires from server using [the questionnaire sync API](https://api.simple.org/api-docs/index.html#tag/Questionnaires). Server responds with all layouts compatible with the DSL version supported by the mobile app.
+
+On Mobile database, the primary key for Questionnaires sync resource is questionnaire_type, and not id. _That is a key difference between Questionnaire and other sync resources._ Reason being, mobile needs only 1 layout per type of a questionnaire. Whenever a new questionnaire of an existing type is created, server sends the updated ID & layout in the sync response, and mobile updates the row for corresponding type.
+
+| questionnaire_type <PK> | id     | layout   |
+|-------------------------|--------|----------|
+| screening_reports       | uuid-1 | {form-1} |
+| supplies_reports        | uuid-2 | {form-2} |
+
+### 2. Mobile _syncs_ questionnaire_responses from server to device
+User inputs for a dynamic form are stored in a questionnaire response. Mobile requests them from server using [this sync API](https://api.simple.org/api-docs/index.html#tag/Questionnaire-Responses).
+
+Other sync resources have explicit keys for recording input (Blood Pressures have "systolic"/"diastolic", Patients have "gender", etc.). Questionnaire response records all user inputs in a single field `content` of JSON data type.
+
+| id      | questionnaire_id | questionnaire_type | facility_id | user_id | content          |
+|---------|------------------|--------------------|-------------|---------|------------------|
+| uuid-11 | uuid-1           | screening_reports  | Surat       | 123     | {"name": "john"} |
+| uuid-22 | uuid-2           | drug_stock_reports | Bangalore   | 456     | {"name": "doe"}  |
+
+### 3. Displaying questionnaires on home page
+![questionnaires-on-home-page](resources/questionnaires-home-page.png)
+
+The decision to display or hide link to a dynamic form on the home page is based on 3 steps:
+1. Mobile has a questionnaire of given type.
+1. Mobile has questionnaire responses for given type.
+1. Config of a questionnaire is enabled for User's facility, which mobile derives from [the facility sync API](https://api.simple.org/api-docs/index.html#tag/facility).
+
+### 4. Displaying a list of forms
+![questionnaire-responses-list](resources/questionnaire-responses-list.png)
+
+When user clicks on a form type from home page, Mobile queries its table for responses of given questionnaire type. Mobile has some custom logic on ordering and style of the forms:
+1. Mobile orders forms in descending order of months by checking for `content["month_date"]` key.
```

Comment:
What happens if mobile can't find a response for a month_date?

Reply:
> If Mobile cannot find these keys in a questionnaire response, it will crash.

Added this line to ADR.